docs/src/spec/precompiles.md

```diff
@@ -0,0 +1,322 @@
+# Precompiles
+
+Precompiles are EVM functions that are prebuilt and optimized as part of the Fe standard library. There are currently nine precompiles available in Fe. The first four precompiles were defined in the original [Ethereum whitepaper](http://gavwood.com/paper.pdf) (`ec_recover`, `SHA2_256`, `ripemd_160`, `identity)`. Four more were added during the [Byzantium fork](https://ethereum.org/en/history/#byzantium) (`mod_exp`, `ec_add`, `ec_mul` and `ec_pairing`). A final precompile, `blake2f` was added in [EIP-152](https://github.com/ethereum/EIPs/issues/152) during the [Istanbul fork](https://ethereum.org/en/history/#istanbul).
+
+The nine precompiles available in the Fe standard library are:
+
+- [`ec_recover`](#ecrecover)
+- [`SHA2 256`](#sha2256)
+- [`ripemd160`](#Ripemd160)
+- [`identity`](#identity)
+- [`mod_exp`](#modexp)
+- [`ec_add`](#ecadd)
+- [`ec_mul`](#ecmul)
+- [`ec_pairing`](#ecpairing)
+- [`blake2f`](#blake2f)
+
+These precompiles are imported as follows:
+
+```fe,ignore,ignore
+use std::precompiles
+```
+
+## `ec_recover`
+
+`ec_recover` is a cryptographic function that retrieves a signer's address from a signed message. It is the fundamental operation used for verifying signatures in Ethereum. Ethereum uses the  Elliptic Curve Digital Signature Algorithm (ECDSA) for verifying signatures. This algorithm uses two parameters, `r` and `s`. Ethereum's implementation also uses an additional 'recovery identifier' parameter, `v`, which is used to identify the correct elliptic curve point from those that can be calculated from `r` and `s` alone.
+
+
+### Parameters
+
+- `hash`: the hash of the signed message, `u256`
+- `v`: the recovery identifier, a number in the range 27-30, `u256`  
+- `r`: elliptic curve parameter, `u256`
+- `s`: elliptic curve parameter, `u256`
+
+### Returns
+
+`ec_recover` returns an address.
+
+### Function signature
+
+```fe,ignore,ignore
+pub fn ec_recover(hash: u256, v: u256, r: u256, s: u256) -> address {}
+```
+
+### Example
+
+```fe,ignore,ignore
+let result: address = precompiles::ec_recover(
+    hash: 0x456e9aea5e197a1f1af7a3e85a3212fa4049a3ba34c2289b4c860fc0b0c64ef3,
+    v: 28,
+    r: 0x9242685bf161793cc25603c231bc2f568eb630ea16aa137d2664ac8038825608,
+    s: 0x4f8ae3bd7535248d0bd448298cc2e2071e56992d0774dc340c368ae950852ada
+)
+```
+
+## `SHA2_256`
+
+`SHA2_256` is a hash function. a hash function generates a unique string of characters of fixed length from arbitrary input data.
+
+
+### Parameters
+
+- `input_buf`: a sequence of bytes to hash, `MemoryBuffer`
```

Comment:
The input argument at the callsite is `buf`; `input_buf` is the name of the value used internally in the function (argument labelling scheme was borrowed from swift, btw). The signature should probably be written `sha2_256(buf: MemoryBuffer) -> u256`.

(I'd argue that these single-argument functions should be defined as e.g. `sha2_256(_ input_buf: MemoryBuffer) -> u256`, so they can be called without a label on the arg.)